application/features/email.feature

```diff
@@ -65,3 +65,154 @@ Feature: Email
+
+    Examples:
+      | sendLostKeyEml | u2fUsedDaysAgo | backupUsedDaysAgo | shouldOrNot |
+      | to send        | 4              | 2                 | should NOT  |
+      | to send        | 444            | 222               | should NOT  |
+      | to send        | 444            | 2                 | should      |
+      | NOT to send    | 444            | 2                 | should NOT  |
+
+  Scenario Outline: When to send get backup codes emails
+    Given we are configured <sendGetBackupCodesEml> get backup codes emails
+      And no mfas exist
+      And a user already exists
+      And a u2f mfa option <u2fExistsOrNot>
+      And a totp mfa option <totpExistsOrNot>
+      And a backup code mfa option <backupExistsOrNot>
+    When I check if a get backup codes email should be sent
+    Then I see that a get backup codes email <shouldOrNot> be sent
+
+    Examples:
+      | sendGetBackupCodesEml | u2fExistsOrNot | totpExistsOrNot | backupExistsOrNot | shouldOrNot |
+      | to send               | does NOT exist | does NOT exist  | does NOT exist    | should NOT  |
+      | to send               | does exist     | does NOT exist  | does NOT exist    | should      |
+      | to send               | does NOT exist | does exist      | does NOT exist    | should      |
+      | to send               | does exist     | does exist      | does NOT exist    | should NOT  |
+      | to send               | does exist     | does exist      | does exist        | should NOT  |
+      | to send               | does NOT exist | does exist      | does exist        | should NOT  |
+      | to send               | does exist     | does NOT exist  | does exist        | should NOT  |
+      | to send               | does NOT exist | does NOT exist  | does exist        | should NOT  |
+      | NOT to send           | does exist     | does exist      | does NOT exist    | should NOT  |
+
+  Scenario Outline:  When to send refresh backup codes emails
+    Given we are configured <sendRefreshBackupCodesEml> refresh backup codes emails
+      And no mfas exist
+      And a user already exists
+      And a backup code mfa option <backupExistsOrNot>
+      And there are <backupCodes> backup codes
+    When I check if a refresh backup codes email should be sent
+    Then I see that a refresh backup codes email <shouldOrNot> be sent
+
+    Examples:
+      | sendRefreshBackupCodesEml | backupExistsOrNot | backupCodes | shouldOrNot |
+      | to send                   | does NOT exist    | 0           | should NOT  |
+      | to send                   | does NOT exist    | 0           | should NOT  |
+      | to send                   | does exist        | 3           | should      |
+      | to send                   | does exist        | 4           | should NOT  |
+      | NOT to send               | does exist        | 2           | should NOT  |
+
+  Scenario Outline: When to send mfa option added emails
```

Comment:
These tests related to sending mfa enabled/disabled/added/removed emails seem to expect these emails to be sent under different circumstances than I expected. For example, this test seems to indicate that an mfa-option-added email should only be sent when there were already at least 2 MFA options in place and another is added, rather than when there is at lease 1 MFA option in place and another is added.

Could these tests be simplified to only care about (A) how many verified MFA options were already present and (B) whether we are adding or deleting an MFA option? (Feel free to suggest a different approach, or to help me better understand the current approach.)

On a related note, should we be sending these emails when we add the MFA option, or when we verify it?

Reply:
Here's an example of what I was thinking:
```
  Scenario Outline: When to send mfa option added emails
    Given we are configured <sendMfaOptionAddedEml> mfa option added emails
      And a user already has <numMfasAlready> verified mfa options
      And they <addOrRemove> an mfa option
    When I check if a mfa option added email should be sent
    Then I see that a mfa option added email <shouldOrNot> be sent

    Examples:
      | sendMfaOptionAddedEml | numMfasAlready | addOrRemove | shouldOrNot |
      | to send               | 0              | add         | should NOT  |
      | to send               | 1              | add         | should      |
      | to send               | 2              | add         | should      |
      | to send               | 0              | remove      | should NOT  |
      | to send               | 1              | remove      | should NOT  |
      | to send               | 2              | remove      | should NOT  |
      | NOT to send           | 0              | add         | should NOT  |
      | NOT to send           | 1              | add         | should NOT  |
      | NOT to send           | 2              | add         | should NOT  |
      | NOT to send           | 0              | remove      | should NOT  |
      | NOT to send           | 1              | remove      | should NOT  |
      | NOT to send           | 2              | remove      | should NOT  |
```
You could then even use this same set of examples for all four of these types of emails (mfa added, mfa enabled, mfa removed, mfa disabled) and only have to change the `shouldOrNot` column's values (and the `sendMfaOptionAddedEml` column's title).

Reply:
For now, I'm just going to reword the Scenario and the "And the mfa even type ..." line to clarify that it already happened in the past.